cfg/qt.cfg - Change the Q_D definition to avoid false positives

winterz opened the pull request:
Change the O_D macro definition by moving the 'const' qualifier
to the front of the statement.

Without this change we get false positives "style: Variable 'd' can be
declared as pointer to const [constVariablePointer]"

The current definition follows Qt correctly as we see this
define in the `QtCore/qtclasshelpermacros.h` file:
\#define Q_D(Class) Class##Private * const d = d_func()

but apparently cppcheck doesn't properly recognize that
as a const variable pointer.

winterz committed:
cfg/kde.cfg - add more KDE macros that are widely used

winterz committed:
cfg/kde.cfg - add more KDE macros that are widely used

winterz committed:
Merge branch 'danmar:main' into main

winterz committed:
Add definitions for Q_DECL_DEPRECATED and Q_DECL_DEPRECATED_X

Add support for Qt macros Q_DECL_DEPRECATED and Q_DECL_DEPRECATED_X
for people use the Qt library.

winterz committed:
Merge branch 'danmar:main' into main

winterz committed:
Merge branch 'danmar:main' into main

winterz committed:
Merge branch 'danmar:main' into main

winterz committed:
cfg/qt.cfg - Change the Q_D definition to avoid false positives

Change the O_D macro definition by moving the 'const' qualifier
to the front of the statement.

Without this change we get false positives "style: Variable 'd' can be
declared as pointer to const [constVariablePointer]"

The current definition follows Qt correctly as we see this
define in the `QtCore/qtclasshelpermacros.h` file:
Q_D(Class) Class##Private * const d = d_func()

but apparently cppcheck doesn't properly recognize that
as a const variable pointer.

ludviggunne committed:
fix #13960: False negative: no error when freeing struct member twice (regression) (#7626)

ludviggunne committed:
fix #13961: Additional information for inline suppression in XML output (#7622)

firewave committed:
reduced usage of `os.chdir()` in Python scripts (#6510)

glankk committed:
Fix #13852 (GUI does not finish analysis properly) (#7633)

Create new `QThread`s for each check, as well as for whole-program
analysis, instead of restarting finished threads which is undefined
behavior (see https://doc.qt.io/qt-6/qthread.html#create). Fixes Trac
#13852.

chrchr-github commented:
Can you add a test in `qt.cpp`?

winterz commented:
yes, I can add a test.

however, I think a larger problem is that cppcheck seems to not like  `FooPrivate * const d = d_func()` whereas it seems ok with  `const FooPrivate *d = d_func()` ?  unless something else is happening.

sdegrande committed:
IRC chan is now on libera.chat (#7628)

In cppcheck html report, points to libera.chat rather than freenode.net

ludviggunne committed:
fix #13964: ErrorLogger: Invalid/missing guideline/classification with `--report-type=misra-cpp-2023` (#7624)

danmar commented:
> however, I think a larger problem is that cppcheck seems to not like FooPrivate * const d = d_func() whereas it seems ok with const FooPrivate *d = d_func() ? unless something else is happening.

some simple sample code that demonstrate a problem would be nice. we should like the first code as far as I know.

winterz committed:
Fix #13970 cfg/qt.cfg - add definition for QDate::isValid() (#7630)

Add the bool isValid() signature.
bool isValid(int, int, int) is already there.

autoantwort committed:
Fixes #13962: fi can be nullptr (#7623)

firewave committed:
iwyu.yml: work around missing includes for Fedora [skip ci] (#7625)

firewave committed:
refs #13618 - cleaned up some `--debug` handling and dependencies (#7592)

chrchr-github committed:
Fix #13975 SymbolDatabase: wrongly set function pointer when called constructor is implicitly defined (#7637)

firewave committed:
refs #13914 - apply enforced language for non-project inputs in GUI (#7627)

firewave committed:
cleaned up includes based on include-what-you-use (#7638)

also cleaned up some `externals` includes and the order.

winterz committed:
test/cfg/runtests.sh -  fix

winterz committed:
Merge branch 'danmar:main' into work/winterz/hack_qd

winterz commented:
I wrote a test here but it isn't showing any problems.
must be something else is happening .  I'll look closer.

in my test, I did run into a checkLibraryFunction issue for `qGetPtrHelper` from the Q_DECLARE_PRIVATE macro.
I'm not sure where to look or where to set `qGetPtrHelper` .  can I get a hint?

danmar commented:
> I'm not sure where to look or where to set qGetPtrHelper . can I get a hint?

hmm.. you already identified that it's from Q_DECLARE_PRIVATE

if a simple configuration is provided like:
```
    <function name="qGetPtrHelper">
         <arg nr="1"/>
    </function>
```
does it work better then?

That configuration is not ideal. Cppcheck can not know what value it returns then. However I am not sure that providing the template would help neither the type deduction can be too complex.
